crates/sui-core/src/checkpoints2/mod.rs

```diff
@@ -216,41 +251,220 @@ impl CheckpointBuilder {
     }
 }
 
+impl CheckpointAggregator {
+    fn new(
+        tables: Arc<CheckpointStoreTables>,
+        notify: Arc<Notify>,
+        exit: watch::Receiver<()>,
+        committee: Committee,
+        output: Box<dyn CertifiedCheckpointOutput>,
+    ) -> Self {
+        let current = None;
+        Self {
+            tables,
+            notify,
+            exit,
+            current,
+            committee,
+            output,
+        }
+    }
+
+    async fn run(mut self) {
+        loop {
+            if let Err(e) = self.run_inner().await {
+                error!(
+                    "Error while aggregating checkpoint, will retry in 1s: {:?}",
+                    e
+                );
+                tokio::time::sleep(Duration::from_secs(1)).await;
+                continue;
+            }
+
+            match select(self.exit.changed().boxed(), self.notify.notified().boxed()).await {
+                Either::Left(_) => {
+                    // return on exit signal
+                    return;
+                }
+                Either::Right(_) => {}
+            }
+        }
+    }
+
+    async fn run_inner(&mut self) -> SuiResult {
+        'outer: loop {
+            let current = if let Some(current) = &mut self.current {
+                current
+            } else {
+                let next_to_certify = self
```

Comment:
Please put these into functions. It will be a common api I believe (to get the latest checkpoint cert, and to get the next checkpoint)

Reply:
I am happy to make it into separate function, but I am not sure this directly should be used from outside.
https://github.com/MystenLabs/sui/pull/6111 is the way I think others should interact with checkpoints.

I can setup time to discuss more between all interested parties